packages/firestore/src/core/sync_engine.ts

```diff
@@ -177,28 +177,40 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
    * server. All the subsequent view snapshots or errors are sent to the
    * subscribed handlers. Returns the targetId of the query.
    */
-  listen(query: Query): Promise<TargetId> {
+  async listen(query: Query): Promise<TargetId> {
     this.assertSubscribed('listen()');
-    assert(
-      !this.queryViewsByQuery.has(query),
-      'We already listen to the query: ' + query
-    );
 
-    return this.localStore.allocateQuery(query).then(queryData => {
+    let targetId;
+    let viewSnapshot;
+
+    const queryView = this.queryViewsByQuery.get(query);
+    if (queryView) {
+      // PORTING NOTE: With Mult-Tab Web, it is possible that a query view
+      // already exists when EventManager calls us for the first time. This
+      // happens when the primary tab is already listening to this query on
+      // behalf of another tab and the user of the primary also starts listening
+      // to the query. EventManager will not have an assigned target ID in this
+      // case and calls `listen` to obtain this ID.
+      targetId = queryView.targetId;
+      this.sharedClientState.addLocalQueryTarget(targetId);
+      viewSnapshot = queryView.view.computeInitialSnapshot();
+    } else {
+      const queryData = await this.localStore.allocateQuery(query);
       const status = this.sharedClientState.addLocalQueryTarget(
         queryData.targetId
       );
-      return this.initializeViewAndComputeInitialSnapshot(
+      targetId = queryData.targetId;
+      viewSnapshot = await this.initializeViewAndComputeInitialSnapshot(
         queryData,
         status === 'current'
-      ).then(viewSnapshot => {
-        if (this.isPrimary) {
-          this.remoteStore.listen(queryData);
-        }
-        this.viewHandler!([viewSnapshot]);
-        return queryData.targetId;
-      });
-    });
+      );
+      if (this.isPrimary) {
+        this.remoteStore.listen(queryData);
```

Comment:
Slightly tangential but since the primary ends up with listens that secondaries don't have I was trying to figure out how we reconcile that during primary transfers.  There's a "TODO: Unlisten from all active targets" comment in this file. Can you change that to TODO(multitab)?  I must have missed it in a previous review.

Reply:
I'm gonna leave this as is for now to avoid merge conflicts with https://github.com/firebase/firebase-js-sdk/pull/994 and https://github.com/firebase/firebase-js-sdk/pull/984/files, both of which address this TODO.